geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+          override def build(): ComplexAttribute = new FeatureImpl(props, ad.asInstanceOf[AttributeDescriptor], id)
+        } else ()=>new Builder() {
+          override def build(): ComplexAttribute = new ComplexAttributeImpl(props, ad.asInstanceOf[AttributeDescriptor], id)
+        }
+
+        abstract class Applier {
+          def apply(d:Decoder, props:JArrayList[Property], b:Builder): Unit
+        }
+
+        val bindings:scala.Array[Applier] = (for((name,st)<-r.fields) yield {
+          lazy val fieldSkipper = new Applier {
+            def apply(d:Decoder,props:JArrayList[Property], b:Builder)=st.skip(d)
+          }
+
+          if(AVRO_NAMESPACE == name.getNamespaceURI) name.getLocalPart match {
+            case "__version__" =>
+              new Applier {
+                def apply(d:Decoder,props:JArrayList[Property],b:Builder)=d.readInt()
+              }
+            case "__fid__" =>
+              new Applier {
+                def apply(d:Decoder,props:JArrayList[Property], b:Builder)={
+                  val s = d.readString()
+                  if(!s.isEmpty) b.id = new FeatureIdImpl(s)
+                }
+              }
+            case "simpleContent" => st match {
+              case n@Nullable(Primitive(t)) =>
+                val st=nearestSimpleType(ct)
+                self.find(st.getBinding, t) match {
+                  case Some(codec) =>
+                    new Applier {
+                      def apply(d:Decoder,props:JArrayList[Property], b:Builder)={
+                        val value = if(n.readNull(d)) null else codec.dec(d)
+                        props add new AttributeImpl(value, new AttributeDescriptorImpl(st.asInstanceOf[AttributeType], GEOT_SIMPLE_CONTENT, 0, 1, true, null), null)
+                      }
+                    }
+                }
+              case _ =>
+                fieldSkipper
+            }
+            case _ =>
+              fieldSkipper
+          } else descByName.get(name) match {
+            case Some(desc) =>
+              val binding = bindDecMany(st, desc.asInstanceOf[AttributeDescriptor])
+              new Applier {
+                def apply(d:Decoder,props:JArrayList[Property], b:Builder)=props.addAll(binding(d))
+              }
+            case _ =>
+              fieldSkipper
+          }
+        }).toArray
+
+        val ret = new DecodeToAttribute{
+          def apply(d: Decoder) = {
+            val b = builderFac()
+            val props = b.props
+            val sz = bindings.length
+            var i = 0
+            while (i < sz) {
+              bindings(i)(d, props, b)
+              i = i + 1
+            }
+            b.build()
+          }
+        }
+        decoders(r)=ret
+        constructing(r)=false
+        ret
+      case _ =>
+        ???
+    }
+
+    private[this] def bindDecMany(s: SchemaType, ad:AttributeDescriptor): DecodeToAttributes = s match {
+      case n@Nullable(st) =>
+        new DecodeToAttributes {

```

Comment:
Consider extracting functions to create each of these.  I think it'd improve readability.